src/promptflow/promptflow/executor/_process_manager.py

```diff
@@ -177,16 +177,28 @@ def end_process(self, i):
         try:
             pid = self._process_info[i].process_id
             process = psutil.Process(pid)
-            process.terminate()
-            process.wait()
-            self._process_info.pop(i)
+            # The subprocess will get terminate signal from input queue, so we need to wait for the process to exit.
+            # If the process is still running after 10 seconds, it will raise psutil.TimeoutExpired exception.
+            process.wait(timeout=10)
         except psutil.NoSuchProcess:
-            bulk_logger.warning(f"Process {pid} had been terminated")
+            bulk_logger.warning(f"Process {pid} had been terminated.")
+        except psutil.TimeoutExpired:
+            try:
+                # If the process is still running after waiting 10 seconds, terminate it.
+                process.terminate()
+                process.wait()
+            except Exception as e:
+                bulk_logger.warning(
+                    f"Unexpected error occurred while end process for index {i} and process id {process.pid}. "
```

Comment:
The message occurs twice, maybe we can have a variable to store it to avoid duplicate.